C/c4CAPI.cc

```diff
@@ -299,6 +307,7 @@ C4Document* c4coll_getDoc(C4Collection *coll,
                           C4DocContentLevel content,
                           C4Error* C4NULLABLE outError) noexcept
 {
+    checkCollectionValid(coll, outError, nullptr);
```

Comment:
Not clear why is this line necessary? Won't coll->getDocument throw if not valid?

Reply:
Again, we cannot throw anything here, this method is marked noexcept!

Reply:
Nevermind, I see that you mean that the following will throw and be caught.  I also want to check for null here though.

Reply:
The cpp counterpart, coll->getDocument(), does not throw, why do we make it error here? 
There are two cases we get nullptr return. A. coll itself is invalid. In this case the underlying method throws, and we report to the C API with error. B. the doc to be looked for does not exist in the collection. In this case why make it error as well? I understand in one case the error is NotOpen and the other case NotFound.

Reply:
I don't know what you are talking about.  I want to check if `coll` is null, to void a SIGSEGV.  I figured if I was doing that much I might as well check early for an invalid collection as well.

Reply:
I see your point: you want to check if coll is null. My impression is that the style of c4 api, such as,
   c4coll_xyz(C4Collection* coll ...) 
is similar to an alias, coll->xyz(...). or use of C pointer for C++ implementation. The assumption is that the caller is responsible for using a non-null pointer. 
I checked c4doc_, c4db_, etc. and almost in all occurrences there is no checking pointer. In fact, in header file, we enclose all declarations with C4_ASSUME_NONNULL_BEGIN/C4_ASSUME_NONNULL_END, meaning all pointers are assumed to be non-null unless otherwise specified.

Reply:
Those assumptions are useless in a language that doesn't understand them, like .NET.  C# has no visibility and so I think it would be better to return back an error that it can use as a .NET exception rather than simply hard crashing.

Reply:
In fact in *our own* deprecated API we make no attempt to check if the default collection is null or not before universally sending it to these APIs.

Reply:
Understood that the compiler directives, such as C4_ASSUME_NONNULL_BEGIN. does not help in runtime, but it is part of APIs, telling the reader (user) how to use the functions (passing the arguments). The vast majority of the functions expect that and therefore don't check for NULL. If there is an unchecked NULL, we should fix the code that passed NULL to the function, rather than reporting error that Platform code would have to mind it with every invocation.

Reply:
Yes, which is exactly why it's useless for bound languages.  What is going to happen is that the program will simply stop and say "exited with code <some meaningless code>" with no information on what went wrong.  This sinks a lot of wasted time trying to figure out what happened in a CI build or a user filed bug report.  You can argue that it is "just a null reference error" but that doesn't convince me because if you do something like `c4coll_XXX(c4db_getDefaultCollection(db, err))` the fact that the collection is null means that it is an invalid collection because it doesn't exist.  I am trying to make the library more easy to debug so we don't just get the equivalent of "my code crashed, here is a non-symbolicated stack trace" as a bug report.  Developers using a managed language are not likely to know the ins and outs of analyzing a native crash.  I do want to add more null checking to other parts of the library as well but I didn't want to weigh down this particular PR with it.  The fact is that without this check, or at least a check in the deprecated API, the test I wrote regarding deleting collections and then setting expiration will outright CRASH.   If you really insist then I can make this a debug build only check, but in my opinion the less we outright crash with signals the less time we will spend going back and forth trying to get context of a stack trace.

Reply:
This is an interesting example. In this case, indeed, a NULL is caused by "default collection not found." Make sense. But let's look at this example,
  c4coll_getDoc(collection, "docID", &error) ==>. error.code == NotFound
Is the document not found in the collection or the collection NULL? 
Now, the platform code check the error,
if (error.code != 0) { Log("document %s is not found in %s", docID, c4coll_getSpec(collection)); // (pseudo code)
oh, crashing, because c4coll_getSpec() does not check NULL.

Reply:
The code in question returns "NotOpen" instead of "NotFound" for the case that the collection is null or invalid.  It's kind of shaky but a null collection is definitely invalid.  So your argument of the overlapping error code is moot.  I'm not a fan of the way our API just says "oh don't worry, an error is thrown down the stack" or "oh don't worry the calling code is always going to honor our annotations" when literally the only way to know the former is to be familiar with the code base, and being wrong about the latter causes hours if not days of debugging.  There is no documentation or comments indicating which collection methods throw and in what situations.  This is how we get ourselves into situations where no one knows what is going on and we get caught by surprises time and time again.  

As for your example, ok let's add a custom null check (can't be the same one because we have tests that make sure that the name and scope are still accessible on an invalid collection) to that API as well.